lib/SILGen/SILGenStmt.cpp

```diff
@@ -356,21 +356,22 @@ void StmtEmitter::visitBraceStmt(BraceStmt *S) {
           continue;
         }
       } else if (auto *E = ESD.dyn_cast<Expr*>()) {
-        // Optional chaining expressions are wrapped in a structure like.
-        //
-        // (optional_evaluation_expr implicit type='T?'
-        //   (call_expr type='T?'
-        //     (exprs...
-        //
-        // Walk through it to find out if the statement is actually implicit.
-        if (auto *OEE = dyn_cast<OptionalEvaluationExpr>(E)) {
-          if (auto *IIO = dyn_cast<InjectIntoOptionalExpr>(OEE->getSubExpr()))
-            if (IIO->getSubExpr()->isImplicit()) continue;
-          if (auto *C = dyn_cast<CallExpr>(OEE->getSubExpr()))
-            if (C->isImplicit()) continue;
-        } else if (E->isImplicit()) {
-          // Ignore all other implicit expressions.
-          continue;
+        if (E->isImplicit()) {
```

Comment:
i made an attempt at unifying the logic across the various existing cases here. i.e. if there's an implicit expression that contains a non-implicit child, then it seems like it should be diagnosed. i'm quite unfamiliar with what the right tool for the job is exactly, and just guessed that `forEachChildExpr` has basically the right semantics (though it did have some disclaimers about not walking into closures, or something along those lines).

if we're not comfortable with this refactor, adding more special cases should also work, though IMO makes the code at bit more convoluted to read.

<details>
<summary>gist of alternative implementation</summary>


```patch
diff --git a/lib/SILGen/SILGenStmt.cpp b/lib/SILGen/SILGenStmt.cpp
index 1babb747197..f27c711bc04 100644
--- a/lib/SILGen/SILGenStmt.cpp
+++ b/lib/SILGen/SILGenStmt.cpp
@@ -365,9 +365,15 @@ void StmtEmitter::visitBraceStmt(BraceStmt *S) {
         // Walk through it to find out if the statement is actually implicit.
         if (auto *OEE = dyn_cast<OptionalEvaluationExpr>(E)) {
           if (auto *IIO = dyn_cast<InjectIntoOptionalExpr>(OEE->getSubExpr()))
-            if (IIO->getSubExpr()->isImplicit()) continue;
+            if (auto *OExE = dyn_cast<OpenExistentialExpr>(IIO->getSubExpr())) {
+              if (OExE->getSubExpr()->isImplicit()) continue;
+            } else if (IIO->getSubExpr()->isImplicit()) {
+              continue;
+            }
           if (auto *C = dyn_cast<CallExpr>(OEE->getSubExpr()))
             if (C->isImplicit()) continue;
+        } else if (auto *OExE = dyn_cast<OpenExistentialExpr>(E)) {
+          if (OExE->getSubExpr() && OExE->getSubExpr()->isImplicit()) continue;
         } else if (E->isImplicit()) {
           // Ignore all other implicit expressions.
           continue;
```
</details>